src/eos-updater-util.c

```diff
@@ -154,14 +211,147 @@ eos_updater_resolve_upgrade (EosUpdater  *updater,
+          if (sv && sv[1])
+            g_hash_table_insert (hw_descriptors, g_strdup ("product_name"),
+                                 g_strdup (g_strstrip (sv[1])));
+        }
+    }
+  else
+    { /* X86 */
+      guint i;
+
+      for (i = 0; dmi_attributes[i]; i++)
+        {
+          gs_free gchar *path = NULL;
+          gs_unref_object GFile *fp = NULL;
+          gs_free gchar *fc = NULL;
+          gsize len;
+
+          path = g_strconcat (DMI_PATH, dmi_attributes[i], NULL);
+          fp = g_file_new_for_path (path);
+          if (g_file_load_contents (fp, NULL, &fc, &len, NULL, NULL))
+          {
+              if (len > 128)
+                fc[128] = '\0';
+              g_hash_table_insert (hw_descriptors, g_strdup (dmi_attributes[i]),
+                                   g_strdup (g_strstrip (fc)));
+          }
+        }
+    }
+
+  if (!g_hash_table_lookup (hw_descriptors, "sys_vendor"))
+    g_hash_table_insert (hw_descriptors, g_strdup ("sys_vendor"),
+                         g_strdup ("EOSUNKNOWN"));
+
+  if (!g_hash_table_lookup (hw_descriptors, "product_name"))
+    g_hash_table_insert (hw_descriptors, g_strdup ("product_name"),
+                         g_strdup ("EOSUNKNOWN"));
+
+  vendor = g_strdup (g_hash_table_lookup (hw_descriptors, "sys_vendor"));
+  product = g_strdup (g_hash_table_lookup (hw_descriptors, "product_name"));
+  p_id = g_strconcat (cleanstr (vendor), " ", cleanstr (product), NULL);
+  message ("Product group: %s", p_id);
+
+  /* Build the HTTP URI */
+  g_hash_table_insert (hw_descriptors, g_strdup ("ref"), g_strdup (o_ref));
+  g_hash_table_insert (hw_descriptors, g_strdup ("commit"), g_strdup (booted));
+  query = soup_form_encode_hash (hw_descriptors);
+  uri = g_strconcat (baseurl, "/", BRANCHES_CONFIG_PATH, "?", query, NULL);
+  g_hash_table_destroy (hw_descriptors);
+  message ("Branches configuration URI: %s", uri);
+
+  /* Download branch configuration data */
+  soup = soup_session_new ();
+  msg = soup_message_new ("GET", uri);
+  status = soup_session_send_message (soup, msg);
+  if (!SOUP_STATUS_IS_SUCCESSFUL (status))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_HOST_NOT_FOUND,
+                   "Failed to download branch config data (HTTP %d),"
+                   " cannot upgrade", status);
+      goto out;
+    }
+
+  g_object_get (G_OBJECT (msg), "response-body", &msg_body, NULL);
+  bkf = g_key_file_new ();
+  if (!g_key_file_load_from_data (bkf, msg_body->data, -1,
+                                  G_KEY_FILE_NONE, error))
+    goto out;
+
+  /* Check for product-specific entry */
+  if (g_key_file_has_group (bkf, p_id))
+    {
+      if (g_key_file_get_boolean (bkf, p_id, ON_HOLD_KEY, NULL))
+        {
+          message ("Product is on hold, nothing to upgrade here");
+          ret = TRUE;
+          goto out;
+        }
+      p_ref = g_key_file_get_string (bkf, p_id, OSTREE_REF_KEY, NULL);
+      if (!p_ref)
```

Comment:
Same as above - could probably just use the error that would be returned. You can use `g_prefix_error()` if really needed